File: MATLAB/generateAnimation.m
function [] = generateAnimation(P,dt, strideLength)
figure()
Px1 = [zeros(size(P,1),1) P(:,[1 3 5])];
Py1 = [zeros(size(P,1),1) P(:,[2 4 6])]; 

Px2 = P(:,[3 7 9]);
Py2 = P(:,[4 8 10]); 

Hl=line(Px1(1,:), Py1(1,:)); hold on
H2=line(Px2(1,:), Py2(1,:)); hold on
axis equal
Hl=handle(Hl);
xlim([-0.6 3.6])
ylim([-0.1 1.7])
xlabel('(m)')
ylabel('(m)')

for i = 1:7
for j=1:size(P,1)
    Hl.XData=Px1(j,:);
    Hl.YData=Py1(j,:);
    H2.XData=Px2(j,:);
    H2.YData=Py2(j,:);    
    pause(dt)
end
Px1 = Px1 + strideLength;
Px2 = Px2 + strideLength;

end

File: Python/generateAnimation.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

def generateAnimation(jointPositions_opt, dt, strideLength, nCycles=7):
    
    P = jointPositions_opt
    
    Px1_t = np.zeros((P[0].shape[0], 4))
    Px1_t[:,1] = P[0]
    Px1_t[:,2] = P[2]
    Px1_t[:,3] = P[4]
    
    Py1 = np.zeros((P[0].shape[0], 4))
    Py1[:,1] = P[1]
    Py1[:,2] = P[3]
    Py1[:,3] = P[5]   
    
    Px2_t = np.zeros((P[0].shape[0], 3))
    Px2_t[:,0] = P[2]
    Px2_t[:,1] = P[6]
    Px2_t[:,2] = P[8]
    
    Py2 = np.zeros((P[0].shape[0], 3))
    Py2[:,0] = P[3]
    Py2[:,1] = P[7]
    Py2[:,2] = P[9]
    
    Px1 = np.copy(Px1_t)
    Px2 = np.copy(Px2_t)
    
    for nCycle in range(1, nCycles):
        Px1 = np.concatenate((Px1, Px1_t + nCycle*strideLength), axis=0)
        Px2 = np.concatenate((Px2, Px2_t + nCycle*strideLength), axis=0)
        Py1 = np.concatenate((Py1, Py1), axis=0)
        Py2 = np.concatenate((Py2, Py2), axis=0)        
    
    fig = plt.figure()
    ax = fig.add_subplot(111, autoscale_on=False, 
                         xlim=(-0.6, 3.6), ylim=(-0.1, 1.7))
    
    line, = ax.plot([], [], 'o-', lw=2, color='blue')
    line2, = ax.plot([], [], 'o-', lw=2, color='blue')
    ax.set_xlabel('(m)')
    ax.set_ylabel('(m)')
    time_template = 'time = %.1fs'
    time_text = ax.text(0.05, 0.9, '', transform=ax.transAxes)    
    
    def init():
        line.set_data([], [])
        line2.set_data([], [])
        time_text.set_text('')
        return line, line2, time_text
    
    
    def animate(i):
        thisx = Px1[i, :]
        thisy = Py1[i, :]
        
        thisx2 = Px2[i, :]
        thisy2 = Py2[i, :]
    
        line.set_data(thisx, thisy)
        line2.set_data(thisx2, thisy2)
        time_text.set_text(time_template % (i*dt))
        return line, line2, time_text
    
    anim = animation.FuncAnimation(fig, animate, P[0].shape[0]*nCycles,
                            interval=int(1000/(1/dt)), blit=True, 
                            init_func=init)
    plt.draw()
    
    return anim   
